File: src/test/java/com/planit/pages/ShopPage.java
package com.planit.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

public class ShopPage extends BasePage {

    private static final By CART_LINK = By.cssSelector("a[href*='cart']");

    public ShopPage(WebDriver driver) {
        super(driver);
    }

    private By getBuyButton(String productName) {
        String xpath = String.format(
            "//li[contains(@class,'product') or contains(@class,'product-wrap')]" +
            "[.//*[contains(text(),'%s')]]" +
            "//a[contains(@class,'btn-success') or contains(text(),'Buy')]",
            productName
        );
        return By.xpath(xpath);
    }

    public void addToCart(String productName, int quantity) {
        By buyButton = getBuyButton(productName);
        for (int i = 0; i < quantity; i++) {
            click(buyButton);
        }
    }

    public CartPage goToCart() {
        click(CART_LINK);
        return new CartPage(driver);
    }
}

File: PlanitAutomation/Pages/ShopPage.cs
using Microsoft.Playwright;

namespace PlanitAutomation.Pages;

/// <summary>
/// Represents the Jupiter Toys shop page.
/// Allows adding products to the cart and navigating to the cart.
/// </summary>
public class ShopPage(IPage page) : BasePage(page)
{
    private const string CartLink = "a[href*='cart']";

    /// <summary>
    /// Returns an XPath selector for the Buy button of the given product.
    /// </summary>
    private static string BuyButtonSelector(string productName) =>
        $"//li[contains(@class,'product') or contains(@class,'product-wrap')]" +
        $"[.//*[contains(text(),'{productName}')]]" +
        $"//a[contains(@class,'btn-success') or contains(text(),'Buy')]";

    /// <summary>Clicks the Buy button for <paramref name="productName"/> <paramref name="quantity"/> times.</summary>
    public async Task AddToCartAsync(string productName, int quantity)
    {
        var selector = BuyButtonSelector(productName);
        for (var i = 0; i < quantity; i++)
            await ClickAsync(selector);
    }

    public async Task<CartPage> GoToCartAsync()
    {
        await ClickAsync(CartLink);
        return new CartPage(Page);
    }
}
